gaeguli/pipeline.c

```diff
@@ -475,78 +452,79 @@ gaeguli_pipeline_new_full (GaeguliVideoSource source, const gchar * device,
    * perphas, implement GInitable?
    */
 
-  g_debug ("source: [%d / %s]", source, device);
-  pipeline = g_object_new (GAEGULI_TYPE_PIPELINE, "source", source, "device",
-      device, "resolution", resolution, "framerate", framerate, NULL);
+  g_debug ("source: [%d / %u]", source, node_id);
+  pipeline = g_object_new (GAEGULI_TYPE_PIPELINE, "source", source, "node-id",
+      node_id, "resolution", resolution, "framerate", framerate, NULL);
 
-  return g_steal_pointer (&pipeline);
-}
+  /* Create the shared lock and shared hash table, here */
+  pipeline->lock =
+      gaeguli_shm_mutex_new ("source_provider_lock", pipeline->node_id);
+  if (!pipeline->lock) {
+    syslog (LOG_ERR, "Failed to create shared lock");
+  }
 
-GaeguliPipeline *
-gaeguli_pipeline_new (void)
-{
-  g_autoptr (GaeguliPipeline) pipeline = NULL;
+  /* create Source pipeline here */
+  gaeguli_pipeline_create_source (pipeline, &error);
 
-  pipeline = gaeguli_pipeline_new_full (DEFAULT_VIDEO_SOURCE,
-      DEFAULT_VIDEO_SOURCE_DEVICE, DEFAULT_VIDEO_RESOLUTION,
-      DEFAULT_VIDEO_FRAMERATE);
-
-  return g_steal_pointer (&pipeline);
+  return pipeline;
 }
 
-static void
-gaeguli_pipeline_set_snapshot_tags (GaeguliPipeline * self, GVariant * tags)
+GaeguliPipeline *
+gaeguli_pipeline_new_full (GaeguliVideoSource source, guint node_id,
+    GaeguliVideoResolution resolution, guint framerate)
 {
-  GVariantIter it;
-  gchar *tag_name;
-  GVariant *val_variant;
-  GstTagSetter *tag_setter = GST_TAG_SETTER (self->snapshot_jifmux);
-
-  g_return_if_fail (g_variant_is_of_type (tags, G_VARIANT_TYPE_VARDICT));
-
-  gst_tag_setter_reset_tags (tag_setter);
-
-  g_variant_iter_init (&it, tags);
-  while (g_variant_iter_next (&it, "{sv}", &tag_name, &val_variant)) {
-    GValue val = G_VALUE_INIT;
+  GaeguliPipeline *pipeline = NULL;
+  /* configure the client socket to communicate with provider daemon */
+  int sockfd;
+
+  if ((sockfd =
+          gaeguli_configure_client_socket
+          (DEFAULT_SOURCE_PROVIDER_CLIENT_SOCK_PATH)) < 0) {
+    syslog (LOG_ERR, "Failed to create client socket");
+    return NULL;
+  }
 
-    if (!gst_tag_exists (tag_name)) {
-      g_warning ("Unknown tag %s", tag_name);
-      goto next;
+  /* send IPC msg to provider daemon */
```

Comment:
To me, this architecture with provider and consumer daemons that in turn spawn another child processes seems unnecessarily complicated. I think we don't need `providerd` and `consumerd` middle-men. The Gaeguli API client application that called `gaeguli_pipeline_new_full ()` IS the manager of all child processes belonging to that pipeline.

I think what `gaeguli_pipeline_new_full ()` should do is to launch a child process that runs the video source pipeline (`v4l2src ! decodebin ! pipewiresink`) and retrieve `node_id` of the new `pipewiresink` from this _VSRC Child_ process. Allocating `node_id` by the client application and passing it to _VSRC Child_ is also an option; point is both must know what `node_id` value is.

Subsequently, each `gaeguli_pipeline_add_srt_target ()` call should spawn a new _Consumer Child_ (`pipewiresrc ! encoder ! srtsink`), passing it _VSRC Child_'s `node_id`.

`gaeguli_pipeline_remove_target ()` terminates its respective _Consumer Child_.
`gaeguli_pipeline_stop ()` terminates _VSRC Child_ and all _Consumer Child_ instances.

The client application monitors the state of its children and should some terminate unexpectedly, `GaeguliPipeline` will launch it again (so it has to remember the parameters the child was created with for potential resurrection).

In this setup, I imagine the needs for IPC between `GaeguliPipeline` and its children besides passing the initial parameters will be minimal if any.

@justinjoy Does this look like what you want?

Reply:
Sorry for the late reply, but I agree. It looks like a simple and clear approach.